lib/os/ring_buffer.c

```diff
@@ -21,6 +29,59 @@ struct ring_element {
 	uint32_t  value  :8;  /**< Room for small integral values */
 };
 
+static uint32_t mod(struct ring_buf *buf, uint32_t val)
+{
+	return likely(buf->mask) ? val & buf->mask : val % buf->size;
+}
+
+/* Check if indexes did not progresses to far (too close to 32 bit wrapping).
+ * If yes then reduce all indexes by arbitrary value.
+ */
+static void item_indexes_rewind(struct ring_buf *buf)
+{
+	uint32_t rewind;
+	uint32_t threshold = ring_buf_get_rewind_threshold();
+
+	if (buf->head < threshold) {
+		return;
+	}
+
+	rewind = buf->size * (threshold / buf->size);
+
+	struct k_spinlock l = {};
+	k_spinlock_key_t key = k_spin_lock(&l);
+
+	buf->tail -= rewind;
+	buf->head -= rewind;
+	k_spin_unlock(&l, key);
+}
+
+/* Check if indexes did not progresses to far (too close to 32 bit wrapping).
```

Comment:
Check if indexes did not progress too far (too close to 32-bit wrapping). If so, then rewind all indexes by an arbitrary value. For byte-mode temporary indexes must also be reduced.